swift/iOS/Chat/ChatLayoutController.swift

```diff
@@ -0,0 +1,120 @@
+        guard !messages.isEmpty else { return false }
+        let lastIndexPath = IndexPath(item: 0, section: messages.count - 1)
+        return messagesCollectionView.indexPathsForVisibleItems.contains(lastIndexPath)
+    }
+}
+
+extension ChatLayoutController: MessagesDataSource {
+    func currentSender() -> SenderType {
+        return currentUser
+    }
+
+    func messageForItem(at indexPath: IndexPath, in _: MessagesCollectionView) -> MessageType {
+        return messages[indexPath.section]
+    }
+
+    func numberOfSections(in _: MessagesCollectionView) -> Int {
+        return messages.count
+    }
+}
+
+extension ChatLayoutController: MessagesDisplayDelegate {
+    func backgroundColor(for message: MessageType, at _: IndexPath, in _: MessagesCollectionView) -> UIColor {
+        return isFromCurrentSender(message: message) ? .primaryColor : UIColor(red: 230 / 255,
+                                                                               green: 230 / 255,
+                                                                               blue: 230 / 255,
+                                                                               alpha: 1)
+    }
+
+    func messageStyle(for message: MessageType, at _: IndexPath, in _: MessagesCollectionView) -> MessageStyle {
+        let tail: MessageStyle.TailCorner = isFromCurrentSender(message: message) ? .bottomRight : .bottomLeft
+        return .bubbleTail(tail, .curved)
+    }
+
+    func configureAvatarView(_ avatarView: AvatarView,
+                             for message: MessageType,
+                             at _: IndexPath,
+                             in _: MessagesCollectionView) {
+        let currentUser = message.sender.displayName
+        let avatar = Avatar(image: nil, initials: "\(currentUser.first ?? "A")")
+        avatarView.set(avatar: avatar)
+    }
+}
+
+extension ChatLayoutController: MessageCellDelegate {
+    func cellTopLabelAttributedText(for message: MessageType, at indexPath: IndexPath) -> NSAttributedString? {
+        if indexPath.section % 3 == 0 {
+            return NSAttributedString(string: MessageKitDateFormatter.shared.string(from: message.sentDate),
+                                      attributes: [NSAttributedString.Key.font: UIFont.boldSystemFont(ofSize: 10),
+                                                   NSAttributedString.Key.foregroundColor: UIColor.darkGray])
+        }
+        return nil
+    }
+
+    func messageTopLabelAttributedText(for message: MessageType, at _: IndexPath) -> NSAttributedString? {
+        let name = message.sender.displayName
+        let font = UIFont.preferredFont(forTextStyle: .caption1)
+        return NSAttributedString(string: name,
+                                  attributes: [NSAttributedString.Key.font: font])
+    }
+
+    func messageBottomLabelAttributedText(for message: MessageType, at _: IndexPath) -> NSAttributedString? {
+        let formatter: DateFormatter = {
```

Comment:
I think it is better if we crate the formatter in the initializer and reuse it rather than create a new one each time this method is called,

Also why you need a closure for this? instead of just doing 
```
let formatter = DateFormatter()
formatter.dateStyle = .medium
```